io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -0,0 +1,158 @@
+
+impl Debug for RebuildLog {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Rebuild log: '{}' ({} segments: {} blocks x {})",
+            self.device_name,
+            self.num_segments,
+            self.num_blocks,
+            self.block_len
+        )
+    }
+}
+
+impl RebuildLog {
+    /// Creates a new rebuild log instance for the given device.
+    pub fn new(device_name: &str, num_blocks: u64, block_len: u64) -> Self {
+        let num_segments = div_ceil(num_blocks * block_len, SEGMENT_SIZE);
+        let mut segments = BitVec::new();
+        segments.grow(num_segments as usize, false);
+        Self {
+            device_name: device_name.to_owned(),
+            segments,
+            num_segments,
+            num_blocks,
+            block_len,
+        }
+    }
+
+    /// Logs the given operation, marking the corresponding segment as modified
+    /// in the case of a write operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `_io_type`: IoType of the operation to log.
+    /// * `lbn`: Logical block number.
+    /// * `lbn_cnt`: Number of logical blocks affected by the operation.
+    fn log_op(&mut self, _io_type: IoType, lbn: u64, lbn_cnt: u64) {
+        let start_seg = self.lbn_to_seg(lbn);
+        let end_seg = self.lbn_to_seg(lbn + lbn_cnt);
+        for i in start_seg ..= end_seg {
+            self.segments.set(i, true);
+        }
+    }
+
+    /// Determines if the given logical block was modified by a write-like
+    /// operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `lbn`: Logical block number.
+    fn is_modified(&self, block_offset: u64) -> bool {
+        let seg = self.lbn_to_seg(block_offset);
+        if seg < self.segments.len() {
+            self.segments[seg]
+        } else {
+            true
+        }
+    }
+
+    /// Calculates the index of segment corresponding to the given logical
+    /// block.
+    fn lbn_to_seg(&self, lbn: u64) -> usize {
+        (lbn * self.block_len / SEGMENT_SIZE) as usize
+    }
+
+    /// Counts the total number of modified blocks.
+    pub fn count_modified_blocks(&self) -> u64 {
+        let segs = self.segments.iter().filter(|i| *i).count() as u64;
+        segs * SEGMENT_SIZE / self.block_len
+    }
+}
+
+/// A handle object for a rebuild log instance.
+/// A rebuild log instance allows to have multiple handle to access it.
+#[derive(Clone)]
+pub struct RebuildLogHandle {
+    log: Arc<Mutex<RebuildLog>>,
+}
+
+impl From<RebuildLog> for RebuildLogHandle {
+    fn from(log: RebuildLog) -> Self {
+        Self {
+            log: Arc::new(Mutex::new(log)),
+        }
+    }
+}
+
+impl Debug for RebuildLogHandle {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.log.lock().fmt(f)
+    }
+}
+
+impl RebuildLogHandle {
+    /// Returns the device name of the rebuild log.
+    pub fn device_name(&self) -> String {
+        self.log.lock().device_name.clone()
```

Comment:
In effect, this puts a mutex on all IO which is something you may want to avoid straight from the start. Instead, it perhaps would be a better idea a bit vector per core?

Reply:
Hello mr mayastor :) 
And then we just collect the "partial-logs" from each core when we want to start rebuilding? (At that point the child is open for W so we don't need to keep the log anymore)

Reply:
It doesn't put mutex in I/O path unless a child is faulted and has a rebuild log created.
When all children are healthy, channels do not have any rebuild logs in their log lists, so channel's for_each_rebuild_log() is almost no-op.

When a channel has a rebuild log attached, locking the log isn't that much of problem:
* lock is held only for setting a single bit in a bit vector.
* a child is taken out of I/O path, so we save lots of CPU cycles due to that fact.
* parking_lot::Mutex is used which is spin-lock based. That means it is just a single cmp_exchange for the most likely case of a single child failed.
* Even for two or more children, I don't see a spinlock much of a problem if it's just a bit is set under the lock.

Having bit vector per core is also possible but I don't see much of improvement here performance-wise, but the memory footprint may increase considerably for larger devices. Contrary, we might even see a decrease of performance due to more CPU cache misses (a often overlooked case when optimizing for pure CPU cycles sacrificing memory use).

And what I want is to avoid adding additional complexity to this code path. Rebuild and fault/retire logic is quite subtle and sometimes hard to reason about, so less is more in this case.

Reply:
> Hello mr mayastor :) And then we just collect the "partial-logs" from each core when we want to start rebuilding? (At that point the child is open for W so we don't need to keep the log anymore)

@tiagolobocastro Gotcha.